File: packages/ts/src/utils/hash.ts
/**
 * Hash helpers re-exported from [`@noble/hashes`](https://github.com/paulmillr/noble-hashes).
 *
 * Each function accepts a UTF-8 string or raw bytes and returns the digest as
 * raw bytes, with a `*Hex` companion returning a lowercase `0x`-prefixed hex
 * string. No new cryptography is implemented here — these are thin, ergonomic
 * wrappers so callers don't need to depend on `@noble/hashes` directly or pick
 * the right submodule path.
 */

import { sha256 as nobleSha256 } from "@noble/hashes/sha256";
import { keccak_256 as nobleKeccak256 } from "@noble/hashes/sha3";
import { ripemd160 as nobleRipemd160 } from "@noble/hashes/ripemd160";

/** Input accepted by the hash helpers: a UTF-8 string or raw bytes. */
export type HashInput = string | Uint8Array;

const encoder = new TextEncoder();

/** Coerce a {@link HashInput} into bytes (UTF-8 encoding strings). */
function toBytes(input: HashInput): Uint8Array {
  return typeof input === "string" ? encoder.encode(input) : input;
}

/** Encode bytes to a lowercase `0x`-prefixed hex string. */
export function toHex(bytes: Uint8Array): string {
  let out = "0x";
  for (let i = 0; i < bytes.length; i++) {
    out += bytes[i].toString(16).padStart(2, "0");
  }
  return out;
}

/** SHA-256 digest of `input` as raw bytes. */
export function sha256(input: HashInput): Uint8Array {
  return nobleSha256(toBytes(input));
}

/** SHA-256 digest of `input` as a `0x`-prefixed hex string. */
export function sha256Hex(input: HashInput): string {
  return toHex(sha256(input));
}

/** keccak-256 digest of `input` as raw bytes (the EVM hashing primitive). */
export function keccak256(input: HashInput): Uint8Array {
  return nobleKeccak256(toBytes(input));
}

/** keccak-256 digest of `input` as a `0x`-prefixed hex string. */
export function keccak256Hex(input: HashInput): string {
  return toHex(keccak256(input));
}

/** RIPEMD-160 digest of `input` as raw bytes. */
export function ripemd160(input: HashInput): Uint8Array {
  return nobleRipemd160(toBytes(input));
}

/** RIPEMD-160 digest of `input` as a `0x`-prefixed hex string. */
export function ripemd160Hex(input: HashInput): string {
  return toHex(ripemd160(input));
}

File: packages/py/src/qorsdk/utils/hash.py
"""Hash helpers: SHA-256, keccak-256, and RIPEMD-160.

Each function accepts a UTF-8 string or raw bytes and returns the digest as raw
bytes, with a ``*_hex`` companion returning a lowercase ``0x``-prefixed hex
string. No new cryptography is implemented here — SHA-256/RIPEMD-160 come from
the standard library and keccak-256 from ``pycryptodome`` (a transitive
dependency of cosmpy/bip-utils, always present).
"""

from __future__ import annotations

import hashlib

from Crypto.Hash import keccak as _keccak

#: Input accepted by the hash helpers: a UTF-8 string or raw bytes.
HashInput = str | bytes


def _to_bytes(data: HashInput) -> bytes:
    return data.encode("utf-8") if isinstance(data, str) else data


def to_hex(data: bytes) -> str:
    """Encode bytes to a lowercase ``0x``-prefixed hex string."""
    return "0x" + data.hex()


def sha256(data: HashInput) -> bytes:
    """SHA-256 digest of ``data`` as raw bytes."""
    return hashlib.sha256(_to_bytes(data)).digest()


def sha256_hex(data: HashInput) -> str:
    """SHA-256 digest of ``data`` as a ``0x``-prefixed hex string."""
    return to_hex(sha256(data))


def keccak256(data: HashInput) -> bytes:
    """keccak-256 digest of ``data`` as raw bytes (the EVM hashing primitive)."""
    h = _keccak.new(digest_bits=256)
    h.update(_to_bytes(data))
    return h.digest()


def keccak256_hex(data: HashInput) -> str:
    """keccak-256 digest of ``data`` as a ``0x``-prefixed hex string."""
    return to_hex(keccak256(data))


def ripemd160(data: HashInput) -> bytes:
    """RIPEMD-160 digest of ``data`` as raw bytes."""
    return hashlib.new("ripemd160", _to_bytes(data)).digest()


def ripemd160_hex(data: HashInput) -> str:
    """RIPEMD-160 digest of ``data`` as a ``0x``-prefixed hex string."""
    return to_hex(ripemd160(data))
